### PR TITLE
Allow the driver to perform AWS STS AssumeRole when talking to AWS APIs

### PR DESCRIPTION
**Pull request by solarkennedy**
RFC: I would like to be able to assume into a different role than what the existing instance profile has, and also not use static IAM creds. This new feature allows the driver to assume role based on a `AWS_STS_ROLE` variable, and maintain fresh creds.

I wasn't sure how else to do this without modifying the code. The AWS golang sdk doesn't seem to have the built-in capability without a code change.

Testing done: Compiled and run on environment. Permission errors before, no permission errors after.

If this change is desired (and I'm not doing it wrong), I can add docs to the readme as an alternative way to get creds, update the helm charts for a new variable to be passed in, etc.

### CODE REVIEW

**Comment by wongma7** [comment]
it looks like the go sdk only respects AWS_ACCESS_KEY_ID/AWS_ACCESS_KEY, AWS_SECRET_ACCESS_KEY/AWS_SECRET_KEY, AWS_SESSION_TOKEN

https://github.com/aws/aws-sdk-go/blob/d87d92b4412739c6ef8917979e922fae98b7654c/aws/credentials/env_provider.go

since AWS_STS_ROLE is not a standard environment variable I would prefer an explicit flag like --aws-sts-role or --aws-role-t-assume or something with some nice help text wdyt?

**Comment by solarkennedy** [comment]
Yes, I can try plumbing down `--aws-sts-role` from the cli.

**Comment by solarkennedy** [comment]
Ok, I went with the more explicit `--aws-assume-role`.
At the same time I'm also working on getting CLA approval.

**Comment by solarkennedy** [comment]
Ok @wongma7 , I got my CLA signed. Let me know what I can do next.

File: pkg/cloud/cloud.go
Comment (wongma7) [inline pkg/cloud/cloud.go:234]: I think these creds will expire. Is there a way to use the provider https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/stscreds/#AssumeRoleProvider to ensure sdk autorefershes them

Comment (solarkennedy) [reply pkg/cloud/cloud.go:234]: Whoops, I misread the docs and thought this was the thing that does that.
Thank you. I'll adjust.

Comment (solarkennedy) [reply pkg/cloud/cloud.go:234]: No, re-reading again, I think this code is correct.

Using `NewCredentials` is how the example code says to do it:
```// Create the credentials from AssumeRoleProvider to assume the role
// referenced by the "myRoleARN" ARN.
creds := stscreds.NewCredentials(sess, "myRoleArn")```
https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/stscreds/

And you can see in the sdk code, `NewCredentials` uses that provider for you:
https://github.com/aws/aws-sdk-go/blob/v1.36.7/aws/credentials/stscreds/assume_role_provider.go#L257

Also empirically, my `aws-ebs-csi-driver controller` process running on this branch has been up since dec4 and can still attach/detach.

Comment (solarkennedy) [reply pkg/cloud/cloud.go:239]: Bump. I've rebased and squashed on top of master as well.

**Comment by solarkennedy** [comment]
Bump. It has been a month. Can I have a second round of review?

File: cmd/options/controller_options.go
Comment (ayberk) [inline cmd/options/controller_options.go:37]: Let's rename this to  `AssumeRoleArn`to make the "arn" requirement more explicit.

File: pkg/cloud/cloud.go
Comment (ayberk) [inline pkg/cloud/cloud.go:231]: Similarly let's rename these "assumeRoleArn".

File: pkg/cloud/cloud.go
Comment (ayberk) [reply pkg/cloud/cloud.go:244]: I think you're right:

```
... The Credentials value will attempt to refresh the credentials using the provider when Credentials.Get is called, if the cached credentials are expiring. 
```

File: cmd/options/controller_options.go
Comment (ayberk) [inline cmd/options/controller_options.go:44]: I think AWS is redundant here but I guess we can keep it. But we should definitely have "arn" in this option, eg, `-aws-assume-role-arn`. We should also briefly explain what happens the operator doesn't provide this value since it's optional.

File: pkg/cloud/cloud.go
Comment (ayberk) [inline pkg/cloud/cloud.go:242]: What's the behavior if this string is not a valid ARN? I'd say we should verify and if it's not, we should return an error and fail the startup. 

Can we add a test case for that?

**Review by ayberk — CHANGES_REQUESTED** [review]
We definitely should have "arn" in the names and handle the invalid ARN case.

Helm chart, more documentation etc can be done in a follow-up PR but they're definitely most welcome.

File: cmd/options/controller_options.go
Comment (solarkennedy) [reply cmd/options/controller_options.go:44]: I put a brief something

File: pkg/cloud/cloud.go
Comment (solarkennedy) [reply pkg/cloud/cloud.go:242]: Well... it seems to do nothing until it actually tries to exercise it? (I'm surprised)

    cloud.go:243] Assuming role into ARN foo

Comment (ayberk) [reply pkg/cloud/cloud.go:242]: We should definitely verify it then and fail early. Can we add a check here? 

Ref: https://docs.aws.amazon.com/sdk-for-go/api/aws/arn/#IsARN

Comment (solarkennedy) [reply pkg/cloud/cloud.go:242]: Added

**Comment by ayberk** [comment]
Thanks for incorporating the feedback. Appreciate the contribution! This LGTM but I want to wait for at least one more review before I merge this to make sure I'm not missing anything. 

/lgtm

**Comment by wongma7** [comment]
/approve 

thanks again

**Comment by AndyXiangLi** [comment]
Running external test to verify this change:
/test pull-aws-ebs-csi-driver-external-test-latest

**Comment by ayberk** [comment]
@solarkennedy we might also need to add a test case for invalid arn to pass the coverall check.

File: pkg/cloud/cloud.go
Comment (AndyXiangLi) [inline pkg/cloud/cloud.go:248]: Can we add unit test for this check?

Comment (solarkennedy) [reply pkg/cloud/cloud.go:248]: added something, this one could be a bad idea because it does actually try to create a ec2 session

**Comment by solarkennedy** [comment]
FYI I've hit my timebox and won't be able to justify working on this code more.
Additionally, we've decided not to use this driver.  If you don't think this functionality will be useful to more people, feel free to close this PR if you don't want to maintain this code.

**Comment by dntosas** [comment]
so until this one is merged, we can't utilize something like `kube2iam` to attach roles to pods, right?

**Comment by ayberk** [comment]
Based on my limited understanding of kube2iam, this shouldn't be a blocker for you to use it. From driver's perspective nothing really changes except you're proxying the requests through the kube2iam container. You can just use the annotations and it should work.

**Pull request closed without merging**
